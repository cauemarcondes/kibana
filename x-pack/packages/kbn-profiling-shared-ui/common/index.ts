/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { i18n } from '@kbn/i18n';

export const NOT_AVAILABLE_LABEL = i18n.translate('xpack.profiling.notAvailableLabel', {
  defaultMessage: 'N/A',
});

export enum AddDataTabs {
  Kubernetes = 'kubernetes',
  Docker = 'docker',
  Binary = 'binary',
  Deb = 'deb',
  RPM = 'rpm',
  ElasticAgentIntegration = 'elasticAgentIntegration',
  Symbols = 'symbols',
}

export const PROFILING_FEEDBACK_LINK = 'https://ela.st/profiling-feedback';
