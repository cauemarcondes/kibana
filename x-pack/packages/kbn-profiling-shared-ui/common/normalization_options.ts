/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

export interface NormalizationOptions {
  baselineScale: number;
  baselineTime: number;
  comparisonScale: number;
  comparisonTime: number;
}

export enum ComparisonMode {
  Absolute = 'absolute',
  Relative = 'relative',
}

export enum NormalizationMode {
  Scale = 'scale',
  Time = 'time',
}
