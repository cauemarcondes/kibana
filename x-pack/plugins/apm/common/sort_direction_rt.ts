/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
import * as t from 'io-ts';

export const toSortDirectionRt = t.union([t.literal('asc'), t.literal('desc')]);
export type SortDirection = t.TypeOf<typeof toSortDirectionRt>;
