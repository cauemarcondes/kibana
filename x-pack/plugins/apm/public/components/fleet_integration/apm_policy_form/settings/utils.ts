/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { PackagePolicyValues } from '../typings';

function mergeNewValue(
  oldValues: PackagePolicyValues,
  key: string,
  value?: any
): PackagePolicyValues {
  return { ...oldValues, [key]: { ...oldValues[key], value } };
}

export function handleFormChange({
  values,
  key,
  value,
  validateForm,
}: {
  values: PackagePolicyValues;
  key: string;
  value?: any;
  validateForm: (values: PackagePolicyValues) => boolean;
}) {
  const newValues = mergeNewValue(values, key, value);
  const isValid = validateForm(newValues);
  return { newValues, isValid };
}
