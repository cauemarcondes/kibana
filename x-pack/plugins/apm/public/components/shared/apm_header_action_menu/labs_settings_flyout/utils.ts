/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { toEditableConfig } from '@kbn/advanced-settings-plugin/public';
import { IUiSettingsClient } from '@kbn/core/public';

type EditableConfig = Record<string, ReturnType<typeof toEditableConfig>>;

export function getEditableConfig({
  settingsKeys,
  uiSettings,
}: {
  settingsKeys: string[];
  uiSettings?: IUiSettingsClient;
}) {
  if (!uiSettings) {
    return {};
  }
  const uiSettingsDefinition = uiSettings.getAll();
  const config: EditableConfig = {};

  settingsKeys.forEach((key) => {
    const settingDef = uiSettingsDefinition?.[key];
    if (settingDef) {
      const editableConfig = toEditableConfig({
        def: settingDef,
        name: key,
        value: settingDef.userValue,
        isCustom: uiSettings.isCustom(key),
        isOverridden: uiSettings.isOverridden(key),
      });
      config[key] = editableConfig;
    }
  });
  return config;
}
