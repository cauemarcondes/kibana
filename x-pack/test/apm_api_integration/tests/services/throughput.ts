/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { service, timerange } from '@elastic/apm-generator';
import expect from '@kbn/expect';
import { first, last, mean, meanBy, uniq } from 'lodash';
import moment from 'moment';
import { ENVIRONMENT_ALL } from '../../../../plugins/apm/common/environment_filter_values';
import { isFiniteNumber } from '../../../../plugins/apm/common/utils/is_finite_number';
import { APIReturnType } from '../../../../plugins/apm/public/services/rest/createCallApmApi';
import { PromiseReturnType } from '../../../../plugins/observability/typings/common';
import archives_metadata from '../../common/fixtures/es_archiver/archives_metadata';
import { FtrProviderContext } from '../../common/ftr_provider_context';
import { registry } from '../../common/registry';
import { roundNumber } from '../../utils';

type ThroughputReturn = APIReturnType<'GET /api/apm/services/{serviceName}/throughput'>;

export default function ApiTest({ getService }: FtrProviderContext) {
  const apmApiClient = getService('apmApiClient');
  const traceData = getService('traceData');

  const archiveName = 'apm_synthetic_8.0.0';
  const metadata = archives_metadata[archiveName];

  registry.when(
    'Throughput with statically generated data',
    { config: 'basic', archives: ['apm_8.0.0_empty'] },
    () => {
      const start = new Date('2021-01-01T00:00:00.000Z').getTime();
      const end = new Date('2021-01-01T00:15:00.000Z').getTime() - 1;

      const GO_PROD_RATE = 10;
      const GO_DEV_RATE = 5;
      const JAVA_PROD_RATE = 20;

      before(async () => {
        const serviceGoProdInstance = service('synth-go', 'production', 'go').instance(
          'instance-a'
        );
        const serviceGoDevInstance = service('synth-go', 'development', 'go').instance(
          'instance-b'
        );
        const serviceJavaInstance = service('synth-java', 'production', 'java').instance(
          'instance-c'
        );

        await traceData.index([
          ...timerange(start, end)
            .interval('1s')
            .rate(GO_PROD_RATE)
            .flatMap((timestamp) =>
              serviceGoProdInstance
                .transaction('GET /api/product/list')
                .duration(1000)
                .timestamp(timestamp)
                .serialize()
            ),
          ...timerange(start, end)
            .interval('1s')
            .rate(GO_DEV_RATE)
            .flatMap((timestamp) =>
              serviceGoDevInstance
                .transaction('GET /api/product/:id')
                .duration(1000)
                .timestamp(timestamp)
                .serialize()
            ),
          ...timerange(start, end)
            .interval('1s')
            .rate(JAVA_PROD_RATE)
            .flatMap((timestamp) =>
              serviceJavaInstance
                .transaction('POST /api/product/buy')
                .duration(1000)
                .timestamp(timestamp)
                .serialize()
            ),
        ]);
      });

      after(() => traceData.clean());

      async function callApi(overrides?: {
        start?: string;
        end?: string;
        transactionType?: string;
        environment?: string;
        kuery?: string;
      }) {
        const response = await apmApiClient.readUser({
          endpoint: 'GET /api/apm/services/{serviceName}/throughput',
          params: {
            path: {
              serviceName: 'synth-go',
            },
            query: {
              start: new Date(start).toISOString(),
              end: new Date(end).toISOString(),
              transactionType: 'request',
              environment: 'production',
              kuery: 'processor.event:transaction',
              ...overrides,
            },
          },
        });

        return response.body;
      }

      describe('when calling it with the default parameters', () => {
        let body: PromiseReturnType<typeof callApi>;

        before(async () => {
          body = await callApi();
        });

        it('returns the expected throughput', () => {
          const throughputValues = uniq(body.currentPeriod.map((coord) => coord.y));
          expect(throughputValues).to.eql([GO_PROD_RATE]);
        });
      });

      describe('when setting environment to all', () => {
        let body: PromiseReturnType<typeof callApi>;

        before(async () => {
          body = await callApi({
            environment: ENVIRONMENT_ALL.value,
          });
        });

        it('returns data for all environments', () => {
          const throughputValues = body.currentPeriod.map(({ y }) => y);
          expect(uniq(throughputValues)).to.eql([GO_PROD_RATE + GO_DEV_RATE]);
        });
      });

      describe('when defining a kuery', () => {
        let body: PromiseReturnType<typeof callApi>;

        before(async () => {
          body = await callApi({
            kuery: `processor.event:transaction and transaction.name:"GET /api/product/:id"`,
            environment: ENVIRONMENT_ALL.value,
          });
        });

        it('returns data that matches the kuery', () => {
          const throughputValues = body.currentPeriod.map(({ y }) => y);
          expect(uniq(throughputValues)).to.eql([GO_DEV_RATE]);
        });
      });
    }
  );

  registry.when('Throughput when data is not loaded', { config: 'basic', archives: [] }, () => {
    it('handles the empty state', async () => {
      const response = await apmApiClient.readUser({
        endpoint: 'GET /api/apm/services/{serviceName}/throughput',
        params: {
          path: {
            serviceName: 'opbeans-go',
          },
          query: {
            start: metadata.start,
            end: metadata.end,
            transactionType: 'request',
            environment: 'ENVIRONMENT_ALL',
            kuery: '',
          },
        },
      });

      expect(response.status).to.be(200);
      expect(response.body.currentPeriod.length).to.be(0);
      expect(response.body.previousPeriod.length).to.be(0);
    });
  });

  async function fetchThroughput(processorEvent: 'metric' | 'transaction') {
    const throughputResponse = await apmApiClient.readUser({
      endpoint: 'GET /api/apm/services/{serviceName}/throughput',
      params: {
        path: { serviceName: 'opbeans-go' },
        query: {
          start: metadata.start,
          end: metadata.end,
          transactionType: 'request',
          environment: 'ENVIRONMENT_ALL',
          kuery: `processor.event : "${processorEvent}"`,
        },
      },
    });
    return throughputResponse.body;
  }

  registry.when(
    'Throughput when data is loaded',
    { config: 'basic', archives: [archiveName] },
    () => {
      describe('without comparisons', () => {
        let throughputMetrics: ThroughputReturn;
        let throughputTransactions: ThroughputReturn;

        before(async () => {
          [throughputMetrics, throughputTransactions] = await Promise.all([
            fetchThroughput('metric'),
            fetchThroughput('transaction'),
          ]);
        });

        it('returns some transactions data', () => {
          expect(throughputTransactions.currentPeriod.length).to.be.greaterThan(0);
          const hasData = throughputTransactions.currentPeriod.some(({ y }) => isFiniteNumber(y));
          expect(hasData).to.equal(true);
        });

        it('returns some metrics data', () => {
          expect(throughputMetrics.currentPeriod.length).to.be.greaterThan(0);
          const hasData = throughputMetrics.currentPeriod.some(({ y }) => isFiniteNumber(y));
          expect(hasData).to.equal(true);
        });

        it('has same mean value for metrics and transactions data', () => {
          const transactionsMean = meanBy(throughputTransactions.currentPeriod, 'y');
          const metricsMean = meanBy(throughputMetrics.currentPeriod, 'y');
          expect(transactionsMean).to.equal(metricsMean);
          expectSnapshot(transactionsMean).toMatchInline(`100`);
          expectSnapshot(metricsMean).toMatchInline(`100`);
        });

        it('has a bucket size of 10 seconds for transactions data', () => {
          const firstTimerange = throughputTransactions.currentPeriod[0].x;
          const secondTimerange = throughputTransactions.currentPeriod[1].x;
          const timeIntervalAsSeconds = (secondTimerange - firstTimerange) / 1000;
          expect(timeIntervalAsSeconds).to.equal(10);
        });

        it('has a bucket size of 1 minute for metrics data', () => {
          const firstTimerange = throughputMetrics.currentPeriod[0].x;
          const secondTimerange = throughputMetrics.currentPeriod[1].x;
          const timeIntervalAsMinutes = (secondTimerange - firstTimerange) / 1000 / 60;
          expect(timeIntervalAsMinutes).to.equal(1);
        });
      });

      describe('with comparisons', () => {
        let throughputResponse: ThroughputReturn;

        before(async () => {
          const response = await apmApiClient.readUser({
            endpoint: 'GET /api/apm/services/{serviceName}/throughput',
            params: {
              path: {
                serviceName: 'opbeans-go',
              },
              query: {
                transactionType: 'request',
                start: moment(metadata.end).subtract(15, 'minutes').toISOString(),
                end: metadata.end,
                comparisonStart: metadata.start,
                comparisonEnd: moment(metadata.start).add(15, 'minutes').toISOString(),
                environment: 'ENVIRONMENT_ALL',
                kuery: '',
              },
            },
          });

          throughputResponse = response.body;
        });

        it('returns some data', () => {
          expect(throughputResponse.currentPeriod.length).to.be.greaterThan(0);
          expect(throughputResponse.previousPeriod.length).to.be.greaterThan(0);

          const currentPeriodNonNullDataPoints = throughputResponse.currentPeriod.filter(({ y }) =>
            isFiniteNumber(y)
          );
          const previousPeriodNonNullDataPoints = throughputResponse.previousPeriod.filter(
            ({ y }) => isFiniteNumber(y)
          );

          expect(currentPeriodNonNullDataPoints.length).to.be.greaterThan(0);
          expect(previousPeriodNonNullDataPoints.length).to.be.greaterThan(0);
        });

        it('has same start time for both periods', () => {
          expect(
            new Date(first(throughputResponse.currentPeriod)?.x ?? NaN).toISOString()
          ).to.equal(new Date(first(throughputResponse.previousPeriod)?.x ?? NaN).toISOString());
        });

        it('has same end time for both periods', () => {
          expect(new Date(last(throughputResponse.currentPeriod)?.x ?? NaN).toISOString()).to.equal(
            new Date(last(throughputResponse.previousPeriod)?.x ?? NaN).toISOString()
          );
        });

        it('returns same number of buckets for both periods', () => {
          expect(throughputResponse.currentPeriod.length).to.be(
            throughputResponse.previousPeriod.length
          );
        });

        it('has same mean value for both periods', () => {
          const currentPeriodAvg = mean(throughputResponse.currentPeriod.map((d) => d.y));
          const previousPeriodAvg = mean(throughputResponse.previousPeriod.map((d) => d.y));
          expect(currentPeriodAvg).to.equal(previousPeriodAvg);
          expectSnapshot(roundNumber(currentPeriodAvg)).toMatchInline(`"98.90"`);
          expectSnapshot(roundNumber(previousPeriodAvg)).toMatchInline(`"98.90"`);
        });
      });
    }
  );
}
