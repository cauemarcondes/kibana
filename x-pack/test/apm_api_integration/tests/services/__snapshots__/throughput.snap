// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APM API tests basic apm_8.0.0 Throughput when data is loaded with kql filter to force metrics-based UI has the correct throughput 1`] = `
Array [
  Object {
    "x": 1627973400000,
    "y": 4,
  },
  Object {
    "x": 1627973460000,
    "y": 2,
  },
  Object {
    "x": 1627973520000,
    "y": 3,
  },
  Object {
    "x": 1627973580000,
    "y": 11,
  },
  Object {
    "x": 1627973640000,
    "y": 4,
  },
  Object {
    "x": 1627973700000,
    "y": 6,
  },
  Object {
    "x": 1627973760000,
    "y": 6,
  },
  Object {
    "x": 1627973820000,
    "y": 11,
  },
  Object {
    "x": 1627973880000,
    "y": 10,
  },
  Object {
    "x": 1627973940000,
    "y": 10,
  },
  Object {
    "x": 1627974000000,
    "y": 12,
  },
  Object {
    "x": 1627974060000,
    "y": 13,
  },
  Object {
    "x": 1627974120000,
    "y": 8,
  },
  Object {
    "x": 1627974180000,
    "y": 10,
  },
  Object {
    "x": 1627974240000,
    "y": 13,
  },
  Object {
    "x": 1627974300000,
    "y": 9,
  },
  Object {
    "x": 1627974360000,
    "y": 0,
  },
  Object {
    "x": 1627974420000,
    "y": 4,
  },
  Object {
    "x": 1627974480000,
    "y": 5,
  },
  Object {
    "x": 1627974540000,
    "y": 6,
  },
  Object {
    "x": 1627974600000,
    "y": 8,
  },
  Object {
    "x": 1627974660000,
    "y": 5,
  },
  Object {
    "x": 1627974720000,
    "y": 5,
  },
  Object {
    "x": 1627974780000,
    "y": 7,
  },
  Object {
    "x": 1627974840000,
    "y": 2,
  },
  Object {
    "x": 1627974900000,
    "y": 15,
  },
  Object {
    "x": 1627974960000,
    "y": 3,
  },
  Object {
    "x": 1627975020000,
    "y": 8,
  },
  Object {
    "x": 1627975080000,
    "y": 13,
  },
  Object {
    "x": 1627975140000,
    "y": 11,
  },
  Object {
    "x": 1627975200000,
    "y": 0,
  },
]
`;

exports[`APM API tests basic apm_8.0.0 Throughput when data is loaded with kql filter to force transaction-based UI calculate throughput per minute because of time range has the correct throughput 1`] = `
Array [
  Object {
    "x": 1627973400000,
    "y": 2,
  },
  Object {
    "x": 1627973430000,
    "y": 2,
  },
  Object {
    "x": 1627973460000,
    "y": 2,
  },
  Object {
    "x": 1627973490000,
    "y": 6,
  },
  Object {
    "x": 1627973520000,
    "y": 2,
  },
  Object {
    "x": 1627973550000,
    "y": 16,
  },
  Object {
    "x": 1627973580000,
    "y": 8,
  },
  Object {
    "x": 1627973610000,
    "y": 2,
  },
  Object {
    "x": 1627973640000,
    "y": 2,
  },
  Object {
    "x": 1627973670000,
    "y": 4,
  },
  Object {
    "x": 1627973700000,
    "y": 8,
  },
  Object {
    "x": 1627973730000,
    "y": 6,
  },
  Object {
    "x": 1627973760000,
    "y": 12,
  },
  Object {
    "x": 1627973790000,
    "y": 8,
  },
  Object {
    "x": 1627973820000,
    "y": 10,
  },
  Object {
    "x": 1627973850000,
    "y": 6,
  },
  Object {
    "x": 1627973880000,
    "y": 16,
  },
  Object {
    "x": 1627973910000,
    "y": 10,
  },
  Object {
    "x": 1627973940000,
    "y": 6,
  },
  Object {
    "x": 1627973970000,
    "y": 12,
  },
  Object {
    "x": 1627974000000,
    "y": 16,
  },
  Object {
    "x": 1627974030000,
    "y": 16,
  },
  Object {
    "x": 1627974060000,
    "y": 8,
  },
  Object {
    "x": 1627974090000,
    "y": 4,
  },
  Object {
    "x": 1627974120000,
    "y": 16,
  },
  Object {
    "x": 1627974150000,
    "y": 4,
  },
  Object {
    "x": 1627974180000,
    "y": 14,
  },
  Object {
    "x": 1627974210000,
    "y": 10,
  },
  Object {
    "x": 1627974240000,
    "y": 14,
  },
  Object {
    "x": 1627974270000,
    "y": 6,
  },
  Object {
    "x": 1627974300000,
    "y": 10,
  },
  Object {
    "x": 1627974330000,
    "y": 0,
  },
  Object {
    "x": 1627974360000,
    "y": 0,
  },
  Object {
    "x": 1627974390000,
    "y": 4,
  },
  Object {
    "x": 1627974420000,
    "y": 6,
  },
  Object {
    "x": 1627974450000,
    "y": 6,
  },
  Object {
    "x": 1627974480000,
    "y": 4,
  },
  Object {
    "x": 1627974510000,
    "y": 6,
  },
  Object {
    "x": 1627974540000,
    "y": 6,
  },
  Object {
    "x": 1627974570000,
    "y": 6,
  },
  Object {
    "x": 1627974600000,
    "y": 10,
  },
  Object {
    "x": 1627974630000,
    "y": 2,
  },
  Object {
    "x": 1627974660000,
    "y": 6,
  },
  Object {
    "x": 1627974690000,
    "y": 4,
  },
  Object {
    "x": 1627974720000,
    "y": 8,
  },
  Object {
    "x": 1627974750000,
    "y": 10,
  },
  Object {
    "x": 1627974780000,
    "y": 4,
  },
  Object {
    "x": 1627974810000,
    "y": 2,
  },
  Object {
    "x": 1627974840000,
    "y": 6,
  },
  Object {
    "x": 1627974870000,
    "y": 6,
  },
  Object {
    "x": 1627974900000,
    "y": 18,
  },
  Object {
    "x": 1627974930000,
    "y": 4,
  },
  Object {
    "x": 1627974960000,
    "y": 2,
  },
  Object {
    "x": 1627974990000,
    "y": 8,
  },
  Object {
    "x": 1627975020000,
    "y": 8,
  },
  Object {
    "x": 1627975050000,
    "y": 12,
  },
  Object {
    "x": 1627975080000,
    "y": 18,
  },
  Object {
    "x": 1627975110000,
    "y": 10,
  },
  Object {
    "x": 1627975140000,
    "y": 8,
  },
  Object {
    "x": 1627975170000,
    "y": 12,
  },
  Object {
    "x": 1627975200000,
    "y": 0,
  },
]
`;

exports[`APM API tests basic apm_8.0.0 Throughput when data is loaded with time comparison has the correct throughput in tpm 1`] = `
Object {
  "currentPeriod": Array [
    Object {
      "x": 1627974310000,
      "y": 0,
    },
    Object {
      "x": 1627974320000,
      "y": 0,
    },
    Object {
      "x": 1627974330000,
      "y": 0,
    },
    Object {
      "x": 1627974340000,
      "y": 0,
    },
    Object {
      "x": 1627974350000,
      "y": 0,
    },
    Object {
      "x": 1627974360000,
      "y": 0,
    },
    Object {
      "x": 1627974370000,
      "y": 0,
    },
    Object {
      "x": 1627974380000,
      "y": 0,
    },
    Object {
      "x": 1627974390000,
      "y": 0,
    },
    Object {
      "x": 1627974400000,
      "y": 0,
    },
    Object {
      "x": 1627974410000,
      "y": 12,
    },
    Object {
      "x": 1627974420000,
      "y": 12,
    },
    Object {
      "x": 1627974430000,
      "y": 0,
    },
    Object {
      "x": 1627974440000,
      "y": 6,
    },
    Object {
      "x": 1627974450000,
      "y": 12,
    },
    Object {
      "x": 1627974460000,
      "y": 6,
    },
    Object {
      "x": 1627974470000,
      "y": 0,
    },
    Object {
      "x": 1627974480000,
      "y": 6,
    },
    Object {
      "x": 1627974490000,
      "y": 0,
    },
    Object {
      "x": 1627974500000,
      "y": 6,
    },
    Object {
      "x": 1627974510000,
      "y": 0,
    },
    Object {
      "x": 1627974520000,
      "y": 18,
    },
    Object {
      "x": 1627974530000,
      "y": 0,
    },
    Object {
      "x": 1627974540000,
      "y": 12,
    },
    Object {
      "x": 1627974550000,
      "y": 6,
    },
    Object {
      "x": 1627974560000,
      "y": 0,
    },
    Object {
      "x": 1627974570000,
      "y": 18,
    },
    Object {
      "x": 1627974580000,
      "y": 0,
    },
    Object {
      "x": 1627974590000,
      "y": 0,
    },
    Object {
      "x": 1627974600000,
      "y": 0,
    },
    Object {
      "x": 1627974610000,
      "y": 18,
    },
    Object {
      "x": 1627974620000,
      "y": 12,
    },
    Object {
      "x": 1627974630000,
      "y": 0,
    },
    Object {
      "x": 1627974640000,
      "y": 0,
    },
    Object {
      "x": 1627974650000,
      "y": 6,
    },
    Object {
      "x": 1627974660000,
      "y": 6,
    },
    Object {
      "x": 1627974670000,
      "y": 12,
    },
    Object {
      "x": 1627974680000,
      "y": 0,
    },
    Object {
      "x": 1627974690000,
      "y": 6,
    },
    Object {
      "x": 1627974700000,
      "y": 0,
    },
    Object {
      "x": 1627974710000,
      "y": 6,
    },
    Object {
      "x": 1627974720000,
      "y": 0,
    },
    Object {
      "x": 1627974730000,
      "y": 18,
    },
    Object {
      "x": 1627974740000,
      "y": 6,
    },
    Object {
      "x": 1627974750000,
      "y": 18,
    },
    Object {
      "x": 1627974760000,
      "y": 0,
    },
    Object {
      "x": 1627974770000,
      "y": 12,
    },
    Object {
      "x": 1627974780000,
      "y": 6,
    },
    Object {
      "x": 1627974790000,
      "y": 6,
    },
    Object {
      "x": 1627974800000,
      "y": 0,
    },
    Object {
      "x": 1627974810000,
      "y": 6,
    },
    Object {
      "x": 1627974820000,
      "y": 0,
    },
    Object {
      "x": 1627974830000,
      "y": 0,
    },
    Object {
      "x": 1627974840000,
      "y": 0,
    },
    Object {
      "x": 1627974850000,
      "y": 12,
    },
    Object {
      "x": 1627974860000,
      "y": 6,
    },
    Object {
      "x": 1627974870000,
      "y": 12,
    },
    Object {
      "x": 1627974880000,
      "y": 0,
    },
    Object {
      "x": 1627974890000,
      "y": 6,
    },
    Object {
      "x": 1627974900000,
      "y": 30,
    },
    Object {
      "x": 1627974910000,
      "y": 24,
    },
    Object {
      "x": 1627974920000,
      "y": 0,
    },
    Object {
      "x": 1627974930000,
      "y": 12,
    },
    Object {
      "x": 1627974940000,
      "y": 0,
    },
    Object {
      "x": 1627974950000,
      "y": 0,
    },
    Object {
      "x": 1627974960000,
      "y": 0,
    },
    Object {
      "x": 1627974970000,
      "y": 6,
    },
    Object {
      "x": 1627974980000,
      "y": 0,
    },
    Object {
      "x": 1627974990000,
      "y": 18,
    },
    Object {
      "x": 1627975000000,
      "y": 0,
    },
    Object {
      "x": 1627975010000,
      "y": 6,
    },
    Object {
      "x": 1627975020000,
      "y": 18,
    },
    Object {
      "x": 1627975030000,
      "y": 6,
    },
    Object {
      "x": 1627975040000,
      "y": 0,
    },
    Object {
      "x": 1627975050000,
      "y": 6,
    },
    Object {
      "x": 1627975060000,
      "y": 6,
    },
    Object {
      "x": 1627975070000,
      "y": 24,
    },
    Object {
      "x": 1627975080000,
      "y": 18,
    },
    Object {
      "x": 1627975090000,
      "y": 18,
    },
    Object {
      "x": 1627975100000,
      "y": 18,
    },
    Object {
      "x": 1627975110000,
      "y": 18,
    },
    Object {
      "x": 1627975120000,
      "y": 6,
    },
    Object {
      "x": 1627975130000,
      "y": 6,
    },
    Object {
      "x": 1627975140000,
      "y": 6,
    },
    Object {
      "x": 1627975150000,
      "y": 18,
    },
    Object {
      "x": 1627975160000,
      "y": 0,
    },
    Object {
      "x": 1627975170000,
      "y": 6,
    },
    Object {
      "x": 1627975180000,
      "y": 24,
    },
    Object {
      "x": 1627975190000,
      "y": 6,
    },
    Object {
      "x": 1627975200000,
      "y": 0,
    },
    Object {
      "x": 1627975210000,
      "y": 0,
    },
  ],
  "previousPeriod": Array [
    Object {
      "x": 1627974310000,
      "y": 0,
    },
    Object {
      "x": 1627974320000,
      "y": 6,
    },
    Object {
      "x": 1627974330000,
      "y": 0,
    },
    Object {
      "x": 1627974340000,
      "y": 0,
    },
    Object {
      "x": 1627974350000,
      "y": 6,
    },
    Object {
      "x": 1627974360000,
      "y": 6,
    },
    Object {
      "x": 1627974370000,
      "y": 0,
    },
    Object {
      "x": 1627974380000,
      "y": 0,
    },
    Object {
      "x": 1627974390000,
      "y": 12,
    },
    Object {
      "x": 1627974400000,
      "y": 0,
    },
    Object {
      "x": 1627974410000,
      "y": 6,
    },
    Object {
      "x": 1627974420000,
      "y": 0,
    },
    Object {
      "x": 1627974430000,
      "y": 6,
    },
    Object {
      "x": 1627974440000,
      "y": 0,
    },
    Object {
      "x": 1627974450000,
      "y": 42,
    },
    Object {
      "x": 1627974460000,
      "y": 6,
    },
    Object {
      "x": 1627974470000,
      "y": 0,
    },
    Object {
      "x": 1627974480000,
      "y": 12,
    },
    Object {
      "x": 1627974490000,
      "y": 12,
    },
    Object {
      "x": 1627974500000,
      "y": 0,
    },
    Object {
      "x": 1627974510000,
      "y": 0,
    },
    Object {
      "x": 1627974520000,
      "y": 0,
    },
    Object {
      "x": 1627974530000,
      "y": 6,
    },
    Object {
      "x": 1627974540000,
      "y": 0,
    },
    Object {
      "x": 1627974550000,
      "y": 6,
    },
    Object {
      "x": 1627974560000,
      "y": 0,
    },
    Object {
      "x": 1627974570000,
      "y": 6,
    },
    Object {
      "x": 1627974580000,
      "y": 0,
    },
    Object {
      "x": 1627974590000,
      "y": 6,
    },
    Object {
      "x": 1627974600000,
      "y": 12,
    },
    Object {
      "x": 1627974610000,
      "y": 12,
    },
    Object {
      "x": 1627974620000,
      "y": 0,
    },
    Object {
      "x": 1627974630000,
      "y": 0,
    },
    Object {
      "x": 1627974640000,
      "y": 6,
    },
    Object {
      "x": 1627974650000,
      "y": 12,
    },
    Object {
      "x": 1627974660000,
      "y": 12,
    },
    Object {
      "x": 1627974670000,
      "y": 12,
    },
    Object {
      "x": 1627974680000,
      "y": 12,
    },
    Object {
      "x": 1627974690000,
      "y": 6,
    },
    Object {
      "x": 1627974700000,
      "y": 6,
    },
    Object {
      "x": 1627974710000,
      "y": 12,
    },
    Object {
      "x": 1627974720000,
      "y": 24,
    },
    Object {
      "x": 1627974730000,
      "y": 0,
    },
    Object {
      "x": 1627974740000,
      "y": 6,
    },
    Object {
      "x": 1627974750000,
      "y": 6,
    },
    Object {
      "x": 1627974760000,
      "y": 6,
    },
    Object {
      "x": 1627974770000,
      "y": 6,
    },
    Object {
      "x": 1627974780000,
      "y": 6,
    },
    Object {
      "x": 1627974790000,
      "y": 30,
    },
    Object {
      "x": 1627974800000,
      "y": 12,
    },
    Object {
      "x": 1627974810000,
      "y": 24,
    },
    Object {
      "x": 1627974820000,
      "y": 6,
    },
    Object {
      "x": 1627974830000,
      "y": 0,
    },
    Object {
      "x": 1627974840000,
      "y": 6,
    },
    Object {
      "x": 1627974850000,
      "y": 12,
    },
    Object {
      "x": 1627974860000,
      "y": 0,
    },
    Object {
      "x": 1627974870000,
      "y": 12,
    },
    Object {
      "x": 1627974880000,
      "y": 0,
    },
    Object {
      "x": 1627974890000,
      "y": 24,
    },
    Object {
      "x": 1627974900000,
      "y": 12,
    },
    Object {
      "x": 1627974910000,
      "y": 24,
    },
    Object {
      "x": 1627974920000,
      "y": 12,
    },
    Object {
      "x": 1627974930000,
      "y": 6,
    },
    Object {
      "x": 1627974940000,
      "y": 30,
    },
    Object {
      "x": 1627974950000,
      "y": 12,
    },
    Object {
      "x": 1627974960000,
      "y": 18,
    },
    Object {
      "x": 1627974970000,
      "y": 6,
    },
    Object {
      "x": 1627974980000,
      "y": 0,
    },
    Object {
      "x": 1627974990000,
      "y": 6,
    },
    Object {
      "x": 1627975000000,
      "y": 6,
    },
    Object {
      "x": 1627975010000,
      "y": 0,
    },
    Object {
      "x": 1627975020000,
      "y": 18,
    },
    Object {
      "x": 1627975030000,
      "y": 18,
    },
    Object {
      "x": 1627975040000,
      "y": 12,
    },
    Object {
      "x": 1627975050000,
      "y": 6,
    },
    Object {
      "x": 1627975060000,
      "y": 6,
    },
    Object {
      "x": 1627975070000,
      "y": 0,
    },
    Object {
      "x": 1627975080000,
      "y": 24,
    },
    Object {
      "x": 1627975090000,
      "y": 6,
    },
    Object {
      "x": 1627975100000,
      "y": 12,
    },
    Object {
      "x": 1627975110000,
      "y": 6,
    },
    Object {
      "x": 1627975120000,
      "y": 6,
    },
    Object {
      "x": 1627975130000,
      "y": 18,
    },
    Object {
      "x": 1627975140000,
      "y": 18,
    },
    Object {
      "x": 1627975150000,
      "y": 24,
    },
    Object {
      "x": 1627975160000,
      "y": 0,
    },
    Object {
      "x": 1627975170000,
      "y": 0,
    },
    Object {
      "x": 1627975180000,
      "y": 12,
    },
    Object {
      "x": 1627975190000,
      "y": 6,
    },
    Object {
      "x": 1627975200000,
      "y": 30,
    },
    Object {
      "x": 1627975210000,
      "y": 0,
    },
  ],
}
`;
